tsinfer/algorithm.py

```diff
@@ -86,9 +86,11 @@ def __init__(
         self.encoded_genotypes_size = num_samples
         if genotype_encoding == constants.GenotypeEncoding.ONE_BIT:
             self.encoded_genotypes_size = num_samples // 8 + int((num_samples % 8) != 0)
-        self.genotype_store = np.zeros(
-            max_sites * self.encoded_genotypes_size, dtype=np.uint8
-        )
+            self.genotype_store = np.zeros(
+                max_sites * self.encoded_genotypes_size, dtype=np.uint8
+            )
+        else:
```

Comment:
I don't think it's a good idea to switch between uint8 and int8 - we should use one or the the other consistently.

I think the simplest thing to do is to stick with uint8 internally for the genotype store, but validate on the way in that values are less than <127.

Reply:
If we stick with uint8, we could preserve `-1` values on output by using `.astype(np.int8)`? E.g.

```python
def get_site_genotypes_subset(self, site_id, samples):
        ...
        else:
            for j, u in enumerate(samples):
                # NB missing data (-1) is stored as 255 in the genotype_store
                g[j] = self.genotype_store[start + u].astype(np.int8)

Reply:
Yeah, we do check that already. The problem is when we are doing the standard 8-bit encoding, using np.uint8 for the store, sticking a "-1" in there (which weirdly we seem to be able to do just fine, and it gets stored as 255), then pulling it out again.

Is the `astype` workaround acceptable?

Reply:
(in the latest commit I also added an explicit assert for <=127 for the 8-bit encoding)

Reply:
Ah, we're already casting the output to a int8. Yeah, what you have is good then.